proposals/interceptors-issues-2024-01.md

```diff
@@ -0,0 +1,153 @@
+
+Further down there is a "Future" section containing issues which do not block the move to stable, but could be addressed in the longer term.
+
+## File path portability
+
+One of our design goals with interceptors was for them to conform to the "portability" requirement of source generators. That means we should be able to run the source generators for a project containing interceptors, commit the results to disk, then remove the source generators from the project and be able to build in a variety of environments. Note that this is different from *determinism*--we don't mind if the resulting assembly differs in things like layout, etc. in each build environment, but want it to be the case that if an interceptor declaration functions in one environment, then it also functions in a different environment, assuming all the sources and dependencies are fully available in each environment.
+
+However, we currently require absolute paths for `InterceptsLocationAttribute`, with pathmap substitution performed if present. This doesn't work between local builds on different paths, or across local builds and CI, because only CI will actually pass `/pathmap` during the build. The local build prefers to preserve local paths (e.g. in the PDB) so that the local versions of the sources can be discovered easily during debugging. The different environments will have different answers for what they expect the paths to be, and the InterceptsLocation will have errors in environments which differ from the "generation-time" environment.
+
+Having looked at several strategies to address here, including adjusting when the build system passes `/pathmap` to the compiler, it seems the most straightforward way is to permit relative paths. This is a strategy which is already proven out by `#line` directives.
+
+When a relative path appears in an `[InterceptsLocation]`, we will resolve it relative to the path of the containing file. The relative path used in the attribute must have `\` normalized to `/`, again, for portability.
+
+```cs
+// /project/src/file1.cs
+class Original
+{
+    static void Interceptable() { }
+
+    void M()
+    {
+        Interceptable();
+    }
+}
+
+// /project/obj/generated/file2.g.cs
+class Interceptors
+{
+    [InterceptsLocation("../../src/file1.cs", ...)] // resolves to "/project/src/file1.cs"
+    public static void Interceptor() { }
+}
+```
+
+This means that generator authors will be best advised to produce relative paths using e.g. `Path.GetRelativePath("/project/src/file1.cs", "/project/obj/generated/file2.g.cs")`.
+
+One reason why we haven't added relative path support already is that source generated files haven't had clearly defined paths. Now, there is a need for a generated file to know what path it "would be written out" to, while the text is being generated, before calling `AddSource` with a `hintName`. We may want to add a new public API like the following for this:
```

Comment:
> Now, there is a need for a generated file to know what path it "would be written out" to, while the text is being generate

This feels a bit off to me. It's not that theer is a need to know what the path would be if it's written out, it's that there is a need to have certainty about the path. Essentially that we need to define it such that relative paths will work. To do this we're choosing to define it as the path it would be written out to. It may seem nit picky but I think it's an important distinction.